source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,389 @@
+        normal_entries_[key] = std::move(value);
+      }
+    }
+
+    // Set the entry for the given handle. If the handle is not valid, do nothing.
+    void insert(InlineKey handle, TPtr value) {
+      if (handle.registryId() != registry_.registryId()) {
+        return;
+      }
+
+      ASSERT(handle.inlineId() < registry_.registrationNum());
+      resetInlineMapEntry(handle.inlineId(), std::move(value));
+    }
+
+    // Remove the entry for the given key. If the key is not found, do nothing.
+    void remove(absl::string_view key) {
+      // Compute the hash value for the key and avoid computing it again in the lookup.
+      const size_t hash_value = absl::Hash<absl::string_view>()(key);
+
+      if (auto entry_id = inlineLookup(key, hash_value); entry_id.has_value()) {
+        resetInlineMapEntry(*entry_id);
+      } else {
+        normal_entries_.erase(key);
+      }
+    }
+
+    // Remove the entry for the given handle. If the handle is not valid, do nothing.
+    void remove(InlineKey handle) {
+      if (handle.registryId() != registry_.registryId()) {
+        return;
+      }
+      resetInlineMapEntry(handle.inlineId());
+    }
+
+    void iterate(std::function<bool(absl::string_view, T*)> callback) const {
+      for (const auto& entry : normal_entries_) {
+        if (!callback(entry.first, entry.second.get())) {
+          return;
+        }
+      }
+
+      for (const auto& id : registry_.registrationMap()) {
+        ASSERT(id.second < registry_.registrationNum());
+
+        auto entry = inline_entries_[id.second];
+        if (entry == nullptr) {
+          continue;
+        }
+
+        if (!callback(id.first, entry)) {
+          return;
+        }
+      }
+    }
+
+    uint64_t size() const { return normal_entries_.size() + inline_entries_size_; }
+
+    bool empty() const { return size() == 0; }
+
+    // Get the registry id that the inline map created by.
+    uint64_t registryId() const { return registry_.registryId(); }
+
+    ~InlineMap() {
+      for (uint64_t i = 0; i < registry_.registrationNum(); ++i) {
+        auto entry = inline_entries_[i];
+        if (entry != nullptr) {
+          delete entry;
+        }
+      }
+      memset(inline_entries_, 0, registry_.registrationNum() * sizeof(TPtr));
+    }
+
+  private:
+    friend class InlineMapRegistry;
+
+    static std::unique_ptr<InlineMap> create(InlineMapRegistry& registry) {
+      return std::unique_ptr<InlineMap<T>>(new (
+          (registry.registrationMap().size() * sizeof(InlineMap<T>::TPtr))) InlineMap(registry));
+    }
+
+    InlineMap(InlineMapRegistry& registry) : registry_(registry) {
+      // Initialize the inline entries to nullptr.
+      memset(inline_entries_, 0, registry_.registrationNum() * sizeof(TPtr));
+    }
+
+    T* lookupImpl(absl::string_view key) const {
+      // Compute the hash value for the key and avoid computing it again in the lookup.
+      const size_t hash_value = InlineMapRegistry::Hash()(key);
```

Comment:
I do not understand why you are working with hash values. flat_hash_map should insulate you fom this.

THis makes me feel like you might get a bug if there's a hash collision.

Reply:
The `absl::flat_hash_map` allow users to pre-compute the hash key manually and use the value when doing hash searching.

If there are multiple maps with same key type, and same key need to be searched, this could reduce redundant hash computing.

Reply:
Oh maybe that's an API offered. But I don't think you need multple maps with the same key. Let's clean that up first :)

Reply:
If we searching a registered inline key with normal key value, then we will do two hash searching here. Although this is not recommend.